kratos/utilities/parallel_utilities.h

```diff
@@ -0,0 +1,202 @@
+#pragma once
+
+//    |  /           |
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:		 BSD License
+//					 Kratos default license:
+//kratos/license.txt
+//
+//  Main authors:    Riccardo Rossi
+//
+
+// System includes
+
+// External includes
+
+// Project includes
+#include "includes/define.h"
+#include "utilities/openmp_utils.h"
+
+namespace Kratos {
+namespace Parallel {
+///@addtogroup Kratos
+///@{
+
+/// type definitions
+typedef int ThreadIdType;  // do not expect this to be an int. It is not for
+                           // C++11 parallelism
+
+inline int GetNumThreads() {
+#ifdef _OPENMP
+    return omp_get_max_threads();
+#else
+    return 1;
+#endif
+}
+
+inline ThreadIdType GetThreadHashId() {
+#ifdef _OPENMP
+    return omp_get_thread_num();
+#else
+    return 0;
+#endif
+}
+
+///@name Kratos Globals
+///@{
+// this function implements a parallel for, in which f is a lambda function
+// which captures its parameters once per very time it is called
+template <class InputIt, class UnaryFunction>
+inline void parallel_for(InputIt first, const int n, UnaryFunction &&f) {
+#pragma omp parallel for firstprivate(first)
+    for (int i = 0; i < n; ++i) {
+        f(first + i);
```

Comment:
I think for proper overloading of ref-qualifiers we need to use `std::forward`
```suggestion
        std::forward<UnaryFunction>(f)(first + i);
```